integration/package-operator/zzz_upgrade_test.go

```diff
@@ -0,0 +1,213 @@
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestUpgrade(t *testing.T) {
+	ctx := logr.NewContext(context.Background(), testr.New(t))
+
+	require.NoError(t, deleteExistingPKO(ctx))
+
+	log := logr.FromContextOrDiscard(ctx)
+	pkg := &corev1alpha1.ClusterPackage{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "package-operator",
+			Namespace: "package-operator-system",
+		},
+	}
+
+	log.Info("Installing latest released PKO")
+	require.NoError(t, createAndWaitFromHTTP(ctx, []string{LatestSelfBootstrapJobURL}))
+
+	jobList := &batchv1.JobList{}
+	require.NoError(t, Client.List(
+		ctx, jobList,
+		client.InNamespace("package-operator-system"),
+	))
+	for i := range jobList.Items {
+		require.NoError(t, Waiter.WaitToBeGone(ctx, &jobList.Items[i], func(obj client.Object) (done bool, err error) { return false, nil }, dev.WithTimeout(6*time.Minute)))
+	}
+
+	require.NoError(t, Waiter.WaitForCondition(ctx, pkg, corev1alpha1.PackageAvailable, metav1.ConditionTrue, dev.WithTimeout(5*time.Minute)))
+	log.Info("Latest released PKO is now available")
+
+	cosList := &corev1alpha1.ClusterObjectSetList{}
+	require.NoError(t, Client.List(ctx, cosList, client.MatchingLabels{"package-operator.run/object-deployment": "package-operator"}))
+	require.Len(t, cosList.Items, 1)
+	existingCos := &cosList.Items[0]
+
+	log.Info("Apply self-bootstrap-job.yaml built from sources")
+	require.NoError(t, createAndWaitFromFiles(ctx, []string{filepath.Join("..", "..", "config", "self-bootstrap-job.yaml")}))
+
+	jobList = &batchv1.JobList{}
+	require.NoError(t, Client.List(
+		ctx, jobList,
+		client.InNamespace("package-operator-system"),
+	))
+	for i := range jobList.Items {
+		require.NoError(t, Waiter.WaitToBeGone(ctx, &jobList.Items[i], func(obj client.Object) (done bool, err error) { return false, nil }, dev.WithTimeout(6*time.Minute)))
+	}
+
+	require.NoError(t, Waiter.WaitForCondition(ctx, existingCos, corev1alpha1.ObjectSetArchived, metav1.ConditionTrue, dev.WithTimeout(6*time.Minute)))
+	require.NoError(t, Waiter.WaitForCondition(ctx, pkg, corev1alpha1.PackageProgressing, metav1.ConditionFalse, dev.WithTimeout(6*time.Minute)))
+	require.NoError(t, Waiter.WaitForCondition(ctx, pkg, corev1alpha1.PackageAvailable, metav1.ConditionTrue, dev.WithTimeout(6*time.Minute)))
+}
+
+func deleteExistingPKO(ctx context.Context) error {
+	log := logr.FromContextOrDiscard(ctx)
+
+	cosList := &corev1alpha1.ClusterObjectSetList{}
+	err := Client.List(
+		ctx, cosList,
+		client.MatchingLabels{
+			"package-operator.run/object-deployment": "package-operator",
+		},
+	)
+	if err != nil {
+		return err
+	}
+	for i := range cosList.Items {
+		cos := &cosList.Items[i]
+		log.Info(fmt.Sprintf("Removing finalizers from cluster object set %s: %v", cos.Name, cos.Finalizers))
+		patch := fmt.Sprintf("{\"metadata\":{\"resourceVersion\":\"%s\",\"finalizers\":[]}}", cos.ResourceVersion)
```

Comment:
could be written as a raw string literal to make the patch more readable for humans and thus less error prone:

```suggestion
		patch := fmt.Sprintf(`{"metadata":{"resourceVersion":"%s","finalizers":[]}}`, cos.ResourceVersion)
```